### PR TITLE
PROV: Fix mixup between general and specialized GCM implementations

### PR DESCRIPTION
**Pull request by levitte**
providers/implementations/ciphers/ciphercommon_gcm_hw.c had an AES
specific GCM update function, while
providers/implementations/ciphers/cipher_aria_gcm_hw.c had the more
general implementation.

This moves them around to have the more general implementation in the
common source, and place the AES specialiation where it belongs.

### CODE REVIEW

File: providers/implementations/ciphers/cipher_aes_gcm_hw.c [hunk 1]
```diff
@@ -59,11 +59,76 @@ static int generic_aes_gcm_initkey(PROV_GCM_CTX *ctx, const unsigned char *key,
     return 1;
 }
 
+static int generic_aes_cipher_update(PROV_GCM_CTX *ctx, const unsigned char *in,
```
Comment (t8m) [inline providers/implementations/ciphers/cipher_aes_gcm_hw.c]: Should it be called generic_aes_gcm_cipher_update? And is the generic_ prefix necessary?

Comment (slontis) [reply providers/implementations/ciphers/cipher_aes_gcm_hw.c]: aes_gcm_cipher_update seems like a reasonable name to me..

Comment (levitte) [reply providers/implementations/ciphers/cipher_aes_gcm_hw.c]: That seems inconsistent, considering `generic_aes_gcm_initkey`.  Should that one be renamed as well?

Comment (slontis) [reply providers/implementations/ciphers/cipher_aes_gcm_hw.c]: It is only used by this file - so maybe generic was a bad name.
All the hw platforms have there own name e.g- s390x_aes_gcm_initkey
This is the default name if there is no hw or if the hw is not capable..

**Comment by t8m** [comment]
This replaces #10685

**Comment by levitte** [comment]
> This replaces #10685

Er...  does it?

**Comment by t8m** [comment]
Yes, AFAIK it does because the code depending on AES_GCM_ASM being defined is moved to a file that already indirectly through cipher_aes_gcm.h includes the aes_platform.h.

**Comment by levitte** [comment]
Aha!  I didn't realise that...

**Review by t8m — DISMISSED** [review]
Ping @levitte will you rename the generic_aes functions to drop the generic prefix? If not I give approval anyway although I'd prefer them to be renamed.

**Comment by levitte** [comment]
I've renamed `generic_aes_gcm_initkey` to `aes_gcm_initkey`, since it's only used locally in `providers/implementations/ciphers/cipher_aes_gcm_hw.c`.  `generic_aes_gmc_cipher_update` is a different story, as it *is* generic for most of the different hardware implementations, so I left that alone.

If someone wants to clean away `generic` from all the names, it should be done everywhere, not just in one spot.  That subject for another PR (and someone other than me :wink:)

File: providers/implementations/ciphers/cipher_aes_gcm_hw.c [hunk 2]
```diff
@@ -59,11 +59,76 @@ static int generic_aes_gcm_initkey(PROV_GCM_CTX *ctx, const unsigned char *key,
...
+    return 1;
+}
+
 static const PROV_GCM_HW aes_gcm = {
-    generic_aes_gcm_initkey,
+    aes_gcm_initkey,
     gcm_setiv,
     gcm_aad_update,
-    gcm_cipher_update,
+    generic_aes_gmc_cipher_update,
```
Comment (t8m) [inline providers/implementations/ciphers/cipher_aes_gcm_hw.c]: Typo gmc.

**Review by t8m — CHANGES_REQUESTED** [review]

File: providers/implementations/ciphers/cipher_aes_gcm_hw.c (hunk 2 above)
Comment (levitte) [reply providers/implementations/ciphers/cipher_aes_gcm_hw.c]: Fixed

**Comment by levitte** [comment]
Ping?

**Review by slontis — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by levitte** [comment]
Merged.

7c66ad65f959fa05ad7b95b0c84384d6e63e56f1 PROV: Fix mixup between general and specialized GCM implementations

**Pull request closed without merging**